Add Spring for Apache Pulsar

onobc opened the pull request:
Adds initial support for Spring for Apache Pulsar.

**Work remaining:**
- [x] docs

onobc commented on buildSrc/src/main/java/org/springframework/boot/build/classpath/CheckClasspathForProhibitedDependencies.java:
```diff
@@ -79,6 +79,9 @@ private boolean prohibited(ModuleVersionIdentifier id) {
 		if (group.equals("javax.money")) {
 			return false;
 		}
+		if (group.equals("javax.xml.bind") && id.getName().equals("jaxb-api")) {
```
The Apache Pulsar client lib unfortunately brings this unfortunate baggage w/ it. The Pulsar admin client relies on this as it still codes to the old `javax.*` API. 

I have raised this https://github.com/apache/pulsar/issues/19856 w/ the Pulsar team and it will be address in the `3.0.0` release (they are going straight from `2.11.0 -> 3.0.0` so its not too far out on timeline but I'm not 100% certain the expected date). 

Would it be possible to augment the plugin to allow this dep. only in Pulsar (until 3.0.0 is out)?

wilkinsona replied:
I don't think we can do that. Even in 2.x we prohibit dependencies from a `javax.*` group in favor of the more modern `jakarta.*` equivalents that use the `javax` namespace for their packages. Even if we tolerated the `javax.xml.bind` API dependencies, I think we'd quickly hit problems with the runtime which uses the same coordinates (`org.glassfish.jaxb:jaxb-runtime`) for its `javax` and `jakarta` implementations.

onobc replied:
> I think we'd quickly hit problems with the runtime which uses the same coordinates (org.glassfish.jaxb:jaxb-runtime) for its javax and jakarta implementations.

I don't see `org.glassfish.jaxb:jaxb-runtime` using any `javax` classes:
* https://mvnrepository.com/artifact/org.glassfish.jaxb/jaxb-bom/4.0.2
  * https://mvnrepository.com/artifact/org.glassfish.jaxb/jaxb-runtime/4.0.2
    * https://mvnrepository.com/artifact/org.glassfish.jaxb/jaxb-core/4.0.2

I don't think that will happen as the classes are "isolated" from one another by package name (`javax` / `jakarta`). In effect, the only thing still looking at the `javax` would be the older Pulsar client admin.

onobc replied:
> the javax classes and the runtime for the jakarta classes on the classpath at the same time due to them have the same Maven coordinates.

The runtime has the same coordinates and only a version difference of that coordinates dictates which version it uses (javax/jakarta), is that correct? 

> The API without a runtime to go with it is of limited or perhaps even no use and I don't want users to have to try to straighten out that mess.

Agreed - that would not be good. 
Although, for the Pulsar client, it does shade the older runtime that it uses (it unfortunately did not shade the API). 
So users would not need to enable the older runtime - I think it would be invisible to them. 

I will setup an example app and see what happens.

wilkinsona replied:
> The runtime has the same coordinates and only a version difference of that coordinates dictates which version it uses (javax/jakarta), is that correct?

Correct. 4.0.x is `jakarta.*`.

onobc replied:
Good news, the updates to the Pulsar client were merged and will make it into the `3.0.0-RC2` release. I will update the core to use that version and we should be rid of this violation.

onobc replied:
We are now using Pulsar `3.0.0` and this is no longer an issue.

onobc commented on spring-boot-project/spring-boot-autoconfigure/build.gradle:
```diff
@@ -58,6 +58,11 @@ dependencies {
 	optional("org.apache.httpcomponents.client5:httpclient5")
 	optional("org.apache.kafka:kafka-streams")
+	optional("org.apache.pulsar:pulsar-client-reactive-producer-cache-caffeine") {
+		// spring-pulsar-reactive includes the pulsar-client-reactive-adapter with
```
The Reactive Pulsar client (`pulsar-client-reactive-adapter`) currently has a couple of prohibited deps that we filter out in Spring Pulsar (`spring-pulsar-reactive`). However, the `pulsar-client-reactive-producer-cache-caffeine` has a dep. on the `pulsar-client-reactive-adapter`. 

I am currently looking into splitting this out a bit in the client to avoid these annoying exclusions.

onobc commented on spring-boot-project/spring-boot-starters/spring-boot-starter-pulsar-reactive/build.gradle:
```diff
@@ -0,0 +1,16 @@
+	api("io.projectreactor:reactor-core")
+	api("org.springframework.pulsar:spring-pulsar-reactive")
+	api("org.apache.pulsar:pulsar-client-reactive-producer-cache-caffeine") {
+		// spring-pulsar-reactive includes the pulsar-client-reactive-adapter with
```
Ditto https://github.com/spring-projects/spring-boot/pull/34763/files#r1148396121

onobc commented on spring-boot-project/spring-boot-starters/spring-boot-starter-pulsar/build.gradle:
```diff
@@ -0,0 +1,11 @@
+
+dependencies {
+	api(project(":spring-boot-project:spring-boot-starters:spring-boot-starter"))
+	api("com.github.ben-manes.caffeine:caffeine")
```
There is WIP to shade the fact that the cache mechanism is using caffeine. It is an internal detail and not intended to swapped out.

onobc replied:
This is done now.

onobc commented on spring-boot-project/spring-boot-autoconfigure/build.gradle:
```diff
@@ -58,6 +58,11 @@ dependencies {
 	}
 	optional("org.apache.httpcomponents.client5:httpclient5")
 	optional("org.apache.kafka:kafka-streams")
+	optional("org.apache.pulsar:pulsar-client-reactive-producer-cache-caffeine") {
```
This transitively brings in caffeine. There is WIP to shade the fact that the cache mechanism is using caffeine. It is an internal detail and not intended to swapped out.

onobc replied:
https://github.com/apache/pulsar-client-reactive/pull/124

onobc commented on spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/pulsar/PulsarAutoConfiguration.java:
```diff
@@ -0,0 +1,178 @@
+	public PulsarTemplate<?> pulsarTemplate(PulsarProducerFactory<?> pulsarProducerFactory,
+			ObjectProvider<ProducerInterceptor> interceptorsProvider, SchemaResolver schemaResolver,
+			TopicResolver topicResolver,
+			ObjectProvider<io.micrometer.observation.ObservationRegistry> observationRegistryProvider,
```
You may notice that we are "sneaking in" the reference to `ObservationRegistry`. I do understand why it is a prohibited import as it belongs under actuator auto-configure. However, the Pulsar template/listener use the registry directly (if provided) to record observations. 

Spring Kafka also uses the registry directly, but rather than constructor injection, it does a bean lookup in [afterSingletonsInstantiated()](https://github.com/spring-projects/spring-kafka/blob/4eacaa9dce252357c7ef4a7632430dc362a3fd8f/spring-kafka/src/main/java/org/springframework/kafka/core/KafkaTemplate.java#L446).  We will change Spring Pulsar to do the same thing shortly. Long term I want to explore instead using a filter or some other mechanism so that it does not have a direct dependency on the registry.

onobc replied:
This has been fixed (no more observation registry in the auto-configuration.

onobc commented:
I have also added support for the following:

- Pulsar connection details in auto-config
- Pulsar service connection from testcontainers
- Pulsar docker compose support + service connections from docker compose

These changes are in my fork (based off this same PR branch) in 3 separate commits, queued up and ready to PR once this base PR is merged:
- https://github.com/onobc/spring-boot/tree/cbono-add-pulsar-service-connection

vpavic commented:
Is this expected to target Spring Boot 3.2 as hinted in [this blog post](https://spring.io/blog/2023/03/24/spring-for-apache-pulsar-0-2-0-available-now)? I'm asking because this is still in 3.x milestone and work on 3.2 is well underway.

wilkinsona commented on spring-boot-project/spring-boot-dependencies/build.gradle:
```diff
@@ -1107,6 +1107,24 @@ bom {
+	library("Pulsar", "3.0.0") {
+		group("org.apache.pulsar") {
+			modules = [
+					"pulsar-client",
+					"pulsar-client-admin",
+					"pulsar-client-all"
```
I think this list of modules is incomplete. It should include every module that's released as a single "package".

wilkinsona commented on spring-boot-project/spring-boot-dependencies/build.gradle:
```diff
@@ -1107,6 +1107,24 @@ bom {
+	library("Pulsar Reactive", "0.3.0") {
+		group("org.apache.pulsar") {
+			modules = [
+					"pulsar-client-reactive-adapter",
+					"pulsar-client-reactive-producer-cache-caffeine",
+					"pulsar-client-reactive-producer-cache-caffeine-shaded"
```
I think this list of modules is incomplete. It should include every module that's released as a single "package".

wilkinsona commented on spring-boot-project/spring-boot-docs/src/docs/asciidoc/messaging.adoc:
```diff
@@ -15,15 +15,12 @@ include::messaging/amqp.adoc[]
 include::messaging/websockets.adoc[]
 
 include::messaging/whats-next.adoc[]
-
-
-
-
-
```
This change seems unrelated and may adversely affect the docs' formatting.

wilkinsona commented on spring-boot-project/spring-boot-docs/src/docs/asciidoc/messaging/pulsar.adoc:
```diff
@@ -0,0 +1,144 @@
+https://pulsar.apache.org/[Apache Pulsar] is supported by providing auto-configuration of the {spring-pulsar-docs}[Spring for Apache Pulsar] project.
+
+Spring Boot will auto-configure and register the classic (imperative) Spring Pulsar components when `org.springframework.pulsar:spring-pulsar` is on the classpath.
+It will do the same for the reactive components when both `org.springframework.pulsar:spring-pulsar-reactive` and Project Reactor are on the classpath.
```
Doesn't `spring-pulsar-reactive` pull in Reactor? If so, I don't think there's any need to mention Reactor here.

wilkinsona commented on spring-boot-project/spring-boot-docs/src/docs/asciidoc/messaging/pulsar.adoc:
```diff
@@ -0,0 +1,144 @@
+There are `spring-boot-starter-pulsar` and `spring-boot-starter-pulsar-reactive` "`Starters`" for conveniently collecting the dependencies for imperative and reactive use, respectively.
+
+TIP: The Reactive support is built on top of the imperative support and therefore the reactive starter includes the imperative starter.
+As such, the imperative components are auto-configured and available in both reactive and imperative scenarios.
```
What happens if you use the imperative components in a reactive app? Will the blocking cause a problem?

wilkinsona commented on spring-boot-project/spring-boot-docs/src/docs/asciidoc/messaging/pulsar.adoc:
```diff
@@ -0,0 +1,144 @@
+      auth-params: "{\"privateKey\":\"file:///Users/some-key.json\",\"issuerUrl\":\"https://auth.server.cloud/", \"audience\":\"urn:sn:acme:dev:my-instance"}"
+----
+
+TIP: Using a map is the recommended approach as it is less error-prone and easier to read.
```
I'd just document the map-based approach.

wilkinsona commented on spring-boot-project/spring-boot-docs/src/docs/asciidoc/messaging/pulsar.adoc:
```diff
@@ -0,0 +1,144 @@
+Spring Pulsar's `PulsarAdministration` client is also auto-configured.
+
+By default, the application tries to connect to a local Pulsar instance at `\http://localhost:8080`.
+This can be adjusted by setting the `spring.pulsar.administration.service-url` property to a different value in the form `(http|https)://<host>:<port>`.
```
Please use `configprop:spring.pulsar.administration.service-url[]` as this ensures that the property exists.

wilkinsona commented on spring-boot-project/spring-boot-docs/src/docs/asciidoc/messaging/pulsar.adoc:
```diff
@@ -0,0 +1,144 @@
+When you use the Pulsar starter, Spring Boot will auto-configure and register a `PulsarClient` bean.
+
+By default, the application tries to connect to a local Pulsar instance at `pulsar://localhost:6650`.
+This can be adjusted by setting the `spring.pulsar.client.service-url` property to a different value.
```
Please use `configprop:spring.pulsar.client.service-url[]` as this ensures that the property exists.

wilkinsona commented on spring-boot-project/spring-boot-docs/src/docs/asciidoc/messaging/pulsar.adoc:
```diff
@@ -0,0 +1,144 @@
+You can configure these components by specifying any of the `spring.pulsar.listener.\*` and `spring.pulsar.consumer.*` prefixed application properties.
+
+[[messaging.pulsar.receiving-reactive]]
+=== Receiving a Message (Reactively)
```
Remove the parentheses?

wilkinsona commented on spring-boot-project/spring-boot-starters/spring-boot-starter-pulsar-reactive/build.gradle:
```diff
@@ -0,0 +1,27 @@
+	// pulsar bouncycastle ext includes duplicate bouncycastle classes
+	ignore { name -> name.startsWith("org/bouncycastle/") }
+
+	// spring-pulsar also shades/relocates caffeine to the same location
+	ignore { name -> name.startsWith("org/springframework/pulsar/shade/com/github/benmanes/caffeine/") }
+	ignore { name -> name.startsWith("org/springframework/pulsar/shade/com/google/errorprone/") }
+	ignore { name -> name.startsWith("org/springframework/pulsar/shade/org/checkerframework/") }
```
Are these still needed? The latest snapshots don't seem to have any shaded classes.

onobc replied:
We do use the shaded version of the reactive cache https://github.com/spring-projects/spring-pulsar/blob/7f5ced49dc012164037a872933da5b954776437d/spring-pulsar-reactive/spring-pulsar-reactive.gradle#L13

Once I fix this https://github.com/apache/pulsar-client-reactive/issues/127 then we can remove these prefixes and they will not be considered duplicates.

onobc replied:
https://github.com/spring-projects/spring-boot/pull/37801

wilkinsona commented on spring-boot-tests/spring-boot-smoke-tests/spring-boot-smoke-test-pulsar-reactive/build.gradle:
```diff
@@ -0,0 +1,15 @@
+	id "org.springframework.boot.conventions"
+}
+
+description = "Spring Boot Pulsar smoke test"
```
Add "Reactive" to the description?

wilkinsona commented on spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/pulsar/PulsarAnnotationDrivenConfiguration.java:
```diff
@@ -0,0 +1,105 @@
+ */
+@Configuration(proxyBeanMethods = false)
+@ConditionalOnClass(EnablePulsar.class)
+public class PulsarAnnotationDrivenConfiguration {
```
Package-private

wilkinsona commented on spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/pulsar/PulsarAnnotationDrivenConfiguration.java:
```diff
@@ -0,0 +1,105 @@
+
+	private final PulsarProperties pulsarProperties;
+
+	public PulsarAnnotationDrivenConfiguration(PulsarProperties pulsarProperties) {
```
Package-private

wilkinsona commented on spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/pulsar/ConsumerConfigProperties.java:
```diff
@@ -0,0 +1,540 @@
+ * @author Chris Bono
+ * @since 3.2
+ */
+public class ConsumerConfigProperties {
```
Could this be nested within `PulsarProperties`?

onobc replied:
No, it is currently used by the SCSt Pulsar Binder config properties as well https://github.com/spring-cloud/spring-cloud-stream/blob/25af6a76b04598eb31bf2f84e59f45cee742b8ea/binders/pulsar-binder/spring-cloud-stream-binder-pulsar/src/main/java/org/springframework/cloud/stream/binder/pulsar/properties/PulsarBinderConfigurationProperties.java#L33-L46

wilkinsona replied:
It shouldn't be doing that. From [the docs](https://docs.spring.io/spring-boot/docs/3.1.2/reference/htmlsingle/#features.external-config.typesafe-configuration-properties.java-bean-binding):

> The properties that map to `@ConfigurationProperties` classes available in Spring Boot, which are configured through properties files, YAML files, environment variables, and other mechanisms, are public API but the accessors (getters/setters) of the class itself are not meant to be used directly.

Please nest it and make the appropriate changes in SCSt Pulsar Binder.

onobc replied:
When the auto-config, binder, and core were in a single repo/project this was not an issue but I figured we would need to decouple it. I will do that and nest as you suggest.

wilkinsona commented on spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/pulsar/ProducerConfigProperties.java:
```diff
@@ -0,0 +1,406 @@
+ * @author Chris Bono
+ * @since 3.2
+ */
+public class ProducerConfigProperties {
```
Could this be nested within `PulsarProperties`?

onobc replied:
No, it is currently used by the SCSt Pulsar Binder config properties as well https://github.com/spring-cloud/spring-cloud-stream/blob/25af6a76b04598eb31bf2f84e59f45cee742b8ea/binders/pulsar-binder/spring-cloud-stream-binder-pulsar/src/main/java/org/springframework/cloud/stream/binder/pulsar/properties/PulsarBinderConfigurationProperties.java#L33-L46

wilkinsona replied:
As above. Please nest it and make the appropriate changes in SCSt Pulsar Binder..

wilkinsona commented on spring-boot-project/spring-boot-autoconfigure/src/main/resources/META-INF/additional-spring-configuration-metadata.json:
```diff
@@ -2133,6 +2133,64 @@
+      "description": "Name of the retry topic where the failing messages will be sent."
+    },
+    {
+      "name": "spring.pulsar.reactive.sender.cache.enabled",
```
Where does this come from? The `Sender` class doesn't have a `Cache`.

onobc replied:
It is hiding in the plethora of other properties ([here](https://github.com/spring-projects/spring-boot/pull/34763/files#diff-9ae54aecfff5dde08e706cda1fda7a45880dcee70f9e3f07f8ab460026d452f8R413) and [defined here](https://github.com/spring-projects/spring-boot/pull/34763/files#diff-9ae54aecfff5dde08e706cda1fda7a45880dcee70f9e3f07f8ab460026d452f8R226)).

Is it best practice to declare the nested props at the top of the containing class? I think so, and by moving the 
```java
private final Cache cache = new Cache();
```
to the top of `Sender` class would make this more readable. Wdyt?

wilkinsona reviewed (changes requested):
This for the PR, @onobc. In addition to the various individual comments that I've made for you consideration, I have the following more general comments that apply in multiple places:

- Please add Kotlin versions of code examples for the docs
- Copyright should be 2012-2023
- All `@since` versions should be 3.2.0
- Some configuration properties could use primitive types (`Integer` -> `int`, `Boolean` -> `boolean`)
- The property classes seem very flat. I think it would be good to group things together more. The four `subscription*` properties in `ConsumerConfigProperties` are one example of this.
- `@Bean` methods and constructors on auto-configuration should be package-private
- Please don't use javadoc `@link` and the like in config property descriptions
- We try to avoid binding properties directly to third-party types such as `org.apache.pulsar.client.api.DeadLetterPolicy`
- Enum properties with default values require additional metadata for those defaults as the annotation processor can't extract them automatically

onobc commented:
Thanks for the detailed review @wilkinsona . From a quick read, the requested changes seem light in nature and all make sense. I will get this updated over the next couple of days.

onobc commented on spring-boot-project/spring-boot-dependencies/build.gradle:
```diff
@@ -1107,6 +1107,96 @@ bom {
 			]
 		}
 	}
+	library("Pulsar", "3.0.0") {
```
Pulsar does not do a good job of dependency hygiene wrt what is server, client, shared and there a TON of published modules. Also, there is not one single arbiter in the repo that says "Here is what I publish" - its scattered about the repo in several different places and I am not clear which is which.

I instead got the list using https://github.com/mthmulders/mcs.

```
mcs search org.apache.pulsar -l 200 | grep ":3.0.0" | sed 's/3.0.0.*/3.0.0/' | sort | pbcopy
```
and
```
mcs search org.apache.pulsar -l 200 | grep ":0.3.0" | sed 's/0.3.0.*/0.3.0/' | pbcopy
```

onobc replied:
Yes, when you have to write a script to figure it out, its quite a lot of modules. I will file an issue and reach out to some committers but I am doubtful we will see any immediate changes in the SB 3.2 timeframe. I will probably end up submitting a PR for it (that would be faster - but have to find the time to do it).

onobc commented:
@wilkinsona I believe I have addressed all of your points except `Please add Kotlin versions of code examples for the docs`. Would you be ok w/ me adding that in a subsequent smaller PR?

onobc commented on spring-boot-project/spring-boot-docs/src/docs/asciidoc/messaging/pulsar.adoc:
```diff
@@ -0,0 +1,132 @@
+
+TIP: The Reactive auto-configuration relies on the imperative auto-configuration and therefore the reactive starter includes the imperative starter.
+As such, the imperative components are auto-configured and available in both reactive and imperative scenarios.
+However, using the imperative components in a fully reactive app is not recommended as it could cause blocking issues.
```
Not as it is currently written. It would be quite the effort to do this. The reactive components leverage the reactive client which adapts the original client (which is imperative and auto-configured). I believe there are other technologies auto-configured in Boot whose reactive components leverage the imperative ones, no?

wilkinsona replied:
> The reactive components leverage the reactive client which adapts the original client (which is imperative and auto-configured)

I wonder if the original, imperative client has to be a bean in the reactive case. Perhaps it does for ease of configuration property binding and the like. Not something to worry about for this PR, but probably worth revisiting down the road before 3.2 GAs.

>  there are other technologies auto-configured in Boot whose reactive components leverage the imperative ones, no?

We have some situations where there's a low-level bean that's shared between imperative and reactive (Elasticsearch client transport and the ElasticsearchClient and ReactiveElasticsearchClient that use it, for example) but I can't think of a situation where we have an imperative bean that we expect applications to use extensively that's then wrapped by a reactive bean. I may well be forgetting something though.

onobc replied:
> I wonder if the original, imperative client has to be a bean in the reactive case. Perhaps it does for ease of configuration property binding and the like. Not something to worry about for this PR, but probably worth revisiting down the road before 3.2 GAs.

I'm happy to take a look and revisit in that timeframe.

onobc commented on spring-boot-tests/spring-boot-smoke-tests/spring-boot-smoke-test-pulsar-reactive/src/test/resources/logback-test.xml:
```diff
@@ -0,0 +1,13 @@
+            <pattern>%d{HH:mm:ss.SSS} [%thread] %-5level %logger - %msg%n</pattern>
+        </encoder>
+    </appender>
+    <root level="WARN">
+        <appender-ref ref="STDOUT"/>
+    </root>
+    <logger name="org.testcontainers" level="WARN"/>
+    <logger name="com.github.dockerjava" level="WARN"/>
+	<logger name="org.springframework.pulsar" level="INFO"/>
```
Good catch - we can remove this now in that case.

onobc commented:
> Thanks for the latest updates, @onobc.

You're welcome @wilkinsona. I will make the updates requested shortly.  

> In addition to my specific comments, I've also spent some time reviewing the configuration properties again. Unfortunately, I'm quite concerned about the sheer number of them.
> 
> Looking at the generated JSON, there are 219 `spring.pulsar` properties. To put that into context, there are 1770 properties in total in `spring-boot-autoconfigure` with 12% of those being for Spring Pulsar. Put another way, adding Spring Pulsar support increases the properties in `spring-boot-autoconfigure` by 14%.
> 
> Maybe it's just the nature of the beast (there are 174 `spring.kafka.*` properties) but I do think it would be worth spending some time to see if the configuration model can be simplified. I think that would be of benefit both to us from a maintenance perspective and to users from an ease-of-use perspective.

I agree, there are many properties. I think it is the nature of the beast though. 
I also agree simplifying would be beneficial. However, I have attempted to do this but do not see a clear-cut path forward.

onobc commented on spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/pulsar/AuthParameterUtils.java:
```diff
@@ -0,0 +1,75 @@
+			.collect(Collectors.toMap((entry) -> convertKebabCaseToCamelCase(entry.getKey()), Map.Entry::getValue));
+	}
+
+	static String maybeConvertToEncodedParamString(Map<String, String> params) {
```
This simplified things nicely.

onobc commented on spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/pulsar/PulsarClientBuilderConfigurer.java:
```diff
@@ -0,0 +1,183 @@
+	private void applyAuthentication(Client clientProperties, ClientBuilder clientBuilder) {
+		if (StringUtils.hasText(clientProperties.getAuthParams())
+				&& !CollectionUtils.isEmpty(clientProperties.getAuthentication())) {
+			throw new IllegalArgumentException(
```
TIL

onobc commented:
@wilkinsona I have covered all of the recent concerns other than:

- too many modules (this will require change in Pulsar)
- too many config props (I don't see a way to reduce this currently)
- don't expose imperative components in reactive app (big effort to do this)
- Add Kotlin examples to docs (would like to do in subsequent PR)

Thanks

philwebb commented:
I find the number of configuration properties quite a concern too. I wonder if we can just add the properties that we think will be popular and lean on `PulsarClientBuilderCustomizer` for the more exotic ones?

One thing we've learnt from previous features is it's easier to add properties later when they are needed than take them away because they aren't.

wilkinsona commented:
We discussed the properties in a team meeting yesterday (without Phil, he was on PTO) and came to the same conclusion. @snicoll reminded my that our Kafka support has got to where it is having evolved over many releases. This makes its large number of properties somewhat easier to justify. Our feeling is that we shouldn't jump straight into such a position with Pulsar and should instead provide plenty of room for things to evolve based on community feedback.

As Phil suggests, the room to evolve can be provided by using one or more customizer callbacks to configure things not covered by configuration properties. It then becomes a question of what should be a configuration property. We'd defer to you on that as we lack the Pulsar expertise to have an informed opinion. As a general rule of thumb, settings that change from environment to environment (dev, staging, prod and so on) are the strongest candidates for configuration properties. Next would be settings that are really commonly tuned. Ideally we'd see a 5-10x reduction in the number of properties that we offer to begin with.

onobc commented:
We took a first pass and were able to knock it down from ~ `196 -> 136` which is about a 30% reduction. Would that work?

We also verified that for all of the properties we are removing, the user has a way to customize them (we are good on that front).

**NOTE** This has not been committed yet - just wanted to run the numbers by the team while the tests are being updated etc..


### Data below 

**NOTE:** the `Admin` and `Client` have `8 + 12` and `30 + 14`, respectively. These are the `regular + ssl` props. I believe the SSL related ones may go away when we move to SSL bundles but I am not 100% sure so I left it like this for now so we can see the count w/ and w/o the SSL props.

| PulsarProperties | Before | After |
| ---------------- |------- | ----- |
| Admin	| 8 + 12 |	8 + 12 |
| Cache	| 3 | 3 |
| Client | 30 + 14 | 7 + 13 |
| Defaults | 1 | 1 |
| Function | 3 | 3 |
| Listener | 6 | 6 |
| Reader | 7 | 7 |
| Template | 1 | 1 |
| ConsumerConfigProps | 36 | 26 |
| ProducerConfigProps | 27 | 17 |
| **Total** | 122 | 79

| ReactivePulsarProperties | Before | After |
| ------------------------ |------- | ----- |
| Cache	| 3 | 3 |
| Consumer | 34 | 27 |
| Listener | 3 | 3 |
| Reader | 8 | 8 |
| Sender | 26 | 16 |
| **Total** | 74 | 57 |

| Original Total | New Total |
| ------------- | ---------- |
| 196 | 136 | 

Which is about a 30% reduction.

onobc commented:
@wilkinsona @philwebb config props reduced ~50% w/ the latest commit. 


### Data below 

**NOTE:** the `Admin` and `Client` have `8 + 12` and `30 + 14`, respectively. These are the `regular + ssl` props. I believe the SSL related ones may go away when we move to SSL bundles but I am not 100% sure so I left it like this for now so we can see the count w/ and w/o the SSL props.

| PulsarProperties | Before | After |
| ---------------- |------- | ----- |
| Admin	| 8 + 12 |	8 + 12 |
| Client | 30 + 14 | 7 + 13 |
| Defaults | 1 | 1 |
| Function | 3 | 3 |
| Listener | 6 | 6 |
| Reader | 7 | 6 |
| Template | 1 | 1 |
| ConsumerConfigProps | 36 | 15 |
| ProducerConfigProps | 27 | 12 |
| **Total** | 119 | 59

| ReactivePulsarProperties | Before | After |
| ------------------------ |------- | ----- |
| Consumer | 34 | 17 |
| Listener | 3 | 3 |
| Reader | 8 | 6 |
| Sender | 26 | 12 |
| **Total** | 71 | 38 |

| Original Total | New Total | % Reduction |
| ------------- | ---------- | ---- |
| 190 | 97 | 51.05% |

onobc commented on spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/pulsar/ClientBuilderSslConfigurer.java:
```diff
@@ -0,0 +1,167 @@
+		}
+	}
+
+	private void applyJksSslProperties(ClientBuilderSslSettings<?> clientBuilderSslSettings, PropertyMapper map,
```
The Pulsar client builder is a bit hokey, in that it has a handful of TLS properties to set when using PEM and a handful of properties to set when using JKS. Then it takes those and under the hood handles the creation of the SSLContext. The SSLContext is not exposed to the builder.
As such I need more info that what is provided by the SSLStoreBundle (basically if I could access the Pem/JksStoreBundleDetails) I would be good). Until that time comes though, I have to dip back down into the SslProperties to get the info I need on the builder.

onobc commented on spring-boot-tests/spring-boot-smoke-tests/spring-boot-smoke-test-pulsar/src/test/java/smoketest/pulsar/SamplePulsarApplicationPemBasedSslTests.java:
```diff
@@ -0,0 +1,62 @@
+import static org.assertj.core.api.Assertions.assertThat;
+
+@SpringBootTest(
+		properties = { "spring.pulsar.client.ssl.bundle=client", "spring.pulsar.administration.ssl.bundle=client",
```
I have added an SSL smoke test pair (JKS+PEM) for both imperative and reactive. 

I see that for data cassandra we always have reactive+ssl  and imperative+ssl smoke test - but none w/o SSL. I think having one w/o SSL is valuable though as not everyone will be using SSL.

Wdyt?

philwebb replied:
We've decided to leave SSL out for now, although we'll document how to set it up using a customizer.

onobc commented on spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/pulsar/PulsarAutoConfiguration.java:
```diff
@@ -78,21 +77,31 @@ public class PulsarAutoConfiguration {
 		List<ProducerBuilderCustomizer<?>> customizers = new ArrayList<>();
 		customizers.add(PulsarPropertyMapper.producerBuilderCustomizer(this.properties));
 		customizers.addAll(customizersProvider.orderedStream().toList());
-		List<ProducerBuilderCustomizer<Object>> lambdaSafeCustomizers = List
```
@philwebb I broke this out into 2 separate beans for consistency w/ the reactive side and to have more concrete bean return types.

onobc commented on spring-boot-project/spring-boot-autoconfigure/src/test/java/org/springframework/boot/autoconfigure/pulsar/PulsarAutoConfigurationTests.java:
```diff
@@ -179,6 +216,49 @@ void injectsExpectedBeans() {
+						.getField(producerFactory, "defaultConfigCustomizers");
+					ProducerBuilder<Object> builder = mock(ProducerBuilder.class);
+					customizersOnFactory.forEach((customizer) -> customizer.customize(builder));
+					// Order of customizers should be default props customizer
```
I added this "customizer execution order" test for each component. The only way I found to do so was via the customizers all executing the same method w/ different values and verify the order. 

I am looking at creating a `CompositeCustomizer` that takes in a default + list of user defined customizers. That way we could test that functionality once and then just make sure that the composite is invoked.

philwebb replied:
I've added a test utility to extract the common testing logic. I don't think we need a `CompositeCustomizer` in order to merge.

onobc commented on spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/pulsar/PulsarProperties.java:
```diff
@@ -535,7 +535,7 @@ public static class Consumer {
 		 */
 		@NestedConfigurationProperty
-		private final DeadLetterPolicy deadLetterPolicy = new DeadLetterPolicy();
+		private DeadLetterPolicy deadLetterPolicy;
```
The `new DeadLetterPolicy()` has a zero for max retry count and the builder fails w/ that non-positive value. I have added a consumer path on the Integration test to make sure we pass our defeault props through an actual consumer builder as the mock does not do this validation.

philwebb replied:
I've added an extra assert in our own mapper code as well to catch problems early.

onobc commented on spring-boot-project/spring-boot-docs/src/docs/asciidoc/messaging/pulsar.adoc:
```diff
@@ -7,7 +7,9 @@ It will do the same for the reactive components when `org.springframework.pulsar
 There are `spring-boot-starter-pulsar` and `spring-boot-starter-pulsar-reactive` "`Starters`" for conveniently collecting the dependencies for imperative and reactive use, respectively.
 
-
+TIP: The Reactive support is built on top of the imperative support and therefore the reactive starter includes the imperative starter.
```
I thought it may be helpful to list this fact and tell them how to "hide the non-reactive" components.

philwebb replied:
I removed this initially because the reactive starter doesn't include the imperative starter (although the reactive starter does include imperative dependencies). I've removed it again for now as I think this kind of detail probably belongs in the Spring Pulsar docs rather than ours.
